docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,78 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
+and will keep giving the same error until the **op** is performed and
+applied to the connection through **s2n_async_pkey_op_apply** call.
+
+Note, it is not safe to call multiple functions on the same **conn** or
+**op** objects from 2 different threads at the same time. Doing so will
+produce undefined behaviour. However it is safe to have a call to
+function involvingi only **conn** at the same time with a call to
+function involving only **op**, as 2 objects are not coupled with each
+other. It is also safe to free **conn** or **op** at any moment with
+respective function calls, with the only exception that **conn** cannot
+be freed inside the **s2n_async_pkey_fn** callback.
+
+```c
+typedef int (*s2n_async_pkey_fn)(struct s2n_connection *conn, struct s2n_async_pkey_op *op);
+extern int s2n_config_set_async_pkey_callback(struct s2n_config *config, s2n_async_pkey_fn fn);
+extern int s2n_async_pkey_op_perform(struct s2n_async_pkey_op *op, s2n_cert_private_key *key);
+extern int s2n_async_pkey_op_apply(struct s2n_async_pkey_op *op, struct s2n_connection *conn);
+extern int s2n_async_pkey_op_free(struct s2n_async_pkey_op *op);
+```
+
+- **op** is an opaque object representing private key operation which
+needs to be performed.
+- **key** is a private key used for operation, can be extracted from
+  **conn** through **s2n_connection_get_selected_cert** and
+  **s2n_cert_chain_and_key_get_key** calls.
+
+**s2n_async_pkey_fn** is invoked every time some action involving
+private key is required during **s2n_negotiate**. The **conn** provides
+a pointer to the connection which triggered the callback, the **op** is
+a pointer to an operation to be performed. The callback takes the
+ownership of **op** object and is responsible for freeing the memory for
+it.
+
+**s2n_config_set_async_pkey_callback** sets up the callback to invoke
+for asynchronous private key operations and enables asynchronous mode.
+
+**s2n_async_pkey_op_perform** performs the **op** allowing it to be used
+to resume the handshake through **s2n_async_pkey_op_apply** call. This
+function can be called only once and any subsequent calls will produce a
+failure. It is safe to call from a different thread, as long as no other
+thread is operating on **op**.
+
+**s2n_async_pkey_op_apply** applies the performed **op** to **conn**
+allowing for the next call to **s2n_negotiate** to proceed through
+handshake. The function will fail if it is called from
+**s2n_async_pkey_fn** callback, or if **op** was not performed through
+**s2n_async_pkey_op_perform** call, or if provided **conn** is different
+from the original **conn** which initiated callback for this **op**. The
+function will succeed only once and any subsequent call will result in
+failure for the same **op**.
+
+**s2n_async_pkey_op_free** frees the memory for **op**. Should eventually
+be called for each of the **op** received in **s2n_async_pkey_fn** to
+avoid any memory leaks.
```

Comment:
I think you mentioned earlier that it is implictly freed if `s2n_async_pkey_op_apply` is called? Can't see that being described here

Reply:
I dropped that behavior, as it was a bit more confusing. Now the caller has to free the op even if it was applied.

Reply:
Ok, definitely makes sense!